Add Windows JVM support

JakeWharton opened the pull request:
Half of #1360. Closes #1678.

crowforkotlin committed:
Add Windows Jvm build support

crowforkotlin commented:
@JakeWharton 
1. The error reported above, build/platform-tests (windows-2025) (pull_request), requires merging PR #1677 first to resolve the Windows CL error. Because I'm concerned that the separate PRs for CL changes and support additions might not be approved, I created two PRs for the two Windows changes (one for Windows CL fixes and the other for Windows JVM Library builds and unit testing support).

2. Secondly, I'm seeing an error with the CL jvmTest macOsX64Test on Mac. I checked the actions in the current trunk branch of Zipline and found that some actions are skipped when executing this task, resulting in the CL passing, while others are not skipped and fail. This is odd. I've reproduced the issue locally and found that a large amount of test code in the test case implementation points to this issue. I've analyzed the build artifacts and debugged them, but I've forgotten the specific details. I'll investigate further tomorrow morning.

The specific error reported by jvmTest macOsX64Test on Mac is in this code implementation.
https://github.com/cashapp/zipline/blob/7cf5c0fade23ed65dfff3327eea90da546df2179/zipline/src/jvmTest/kotlin/app/cash/zipline/ADummyTest.kt#L36

JakeWharton commented:
Fixing the error does not require your PR. We should not be running `apiCheck` on each shard of platform-tests because it forces native compilation of all targets.

crowforkotlin commented:
@JakeWharton Thanks for the explanation. I agree that running apiCheck on each platform-tests shard is inefficient and triggers unnecessary native compilations, which is something we should definitely optimize.
However, I believe the direct cause of the windows-2025 check failure is a more fundamental build error. Looking at the logs, the task fails at :zipline:linuxX64Quickjs with a clang error:
code

* What went wrong:
Execution failed for task ':zipline:linuxX64Quickjs'.
> Process 'command 'C:\Users\runneradmin/.cklib/llvm-11.1.0-windows-x64-2/bin/clang'' finished with non-zero exit value 1
This error indicates that the native compilation task itself is broken on the Windows environment, and it will fail whenever it's triggered, regardless of apiCheck. My other PR, #1677, is specifically intended to fix these kinds of Windows command-line (CL) build issues.
So, my understanding is that we first need to merge #1677 to fix the underlying build problem and get the Windows build to a healthy state. After that, we can proceed with optimizing the CI configuration by removing the redundant apiCheck calls to improve efficiency.

JakeWharton commented:
The Windows CI shard should not be triggering Linux compilation. Whether or not the build failure itself should be fixed is an orthogonal problem.

crowforkotlin commented:
Ah, that makes perfect sense. The issue is apiCheck forcing cross-compilation everywhere.

JakeWharton committed:
Add Windows CI runner

JakeWharton committed:
Remove OS from folder path

It is redundant information with our supported platforms, as we can already use the extension to differentiate.

JakeWharton committed:
Remove dedicated Zig file

This information is in the TROUBLESHOOTING.md (for better or worse).

JakeWharton committed:
Use platform-specific line separator in test value

JakeWharton committed:
Change log

JakeWharton committed:
Windows ARM

JakeWharton committed:
JDK 21 downgrade

Only JDK 21 and 25 are available on Windows ARM, and we can't go 25 (yet).

JakeWharton committed:
Simplify version reading

JakeWharton commented on .github/workflows/.java-version:
```diff
@@ -1 +1 @@
-24
+21
```
There's only "LTS" builds for Windows ARM, and we can't jump to 25 yet

JakeWharton commented on .github/workflows/build.yaml:
```diff
@@ -66,6 +66,8 @@ jobs:
           - macos-15
           - ubuntu-24.04
           - ubuntu-24.04-arm
+          - windows-2025
+          - windows-11-arm
```
GitHub just continually crushing it on naming these things...
